### PR TITLE
Add ability to setup an Android Account with XOAUTH2

### PR DESCRIPTION
**Pull request by wiktor-k**
Hello all, this is PR to add XOAUTH2 account support.

The code works and the account is completely usable but I'm looking for feedback on the code quality and UX (that I'm no expert in).

If anyone wants to test this locally they need to add the debug cert's fingerprint here: https://console.cloud.google.com/apis/credentials/oauthclient (choose Android).

A small showcase what is currently implemented:

1. Small "Use GMail Account" button:

screenshot_1545043689

2. Clicking it brings standard android account chooser:

r4ordfttwyidqjwfzlbv3c0yfycilfbrtilsjw4_9ww_c2edc1fa4f jpg small

3. Selecting an account brings Android OAuth permissions dialog:

vwv04umvnienrjzfo6wuhb3exen6ssedhxn67jwehzm_35e0acae8f jpg small

4. After allowing everything goes as with auto setup:

nsuengsq9bacu0tqqqgdkqxkp4alcwmlfaakcyclmdi_c538c60b3f jpg small

2nw2cnohzivjyi6i7xg9ckxzdvqkzfe0hwozafkqveo_267f16a282 jpg small

Emails are there, sending works.

This code heavily reuses old code from @philipwhiuk, I just restructured it a little (I'm planning to clean it up more).

* Follows our existing [codestyle](https://github.com/k9mail/k-9/wiki/CodeStyle). Didn't check yet.
* Does not break any unit tests. :heavy_check_mark: 
* Contains a reference to the issue that it fixes. :heavy_check_mark: 
* For cosmetic changes add one or multiple images, if possible. Attached above.

### CODE REVIEW

File: app/k9mail/src/main/java/com/fsck/k9/backends/ImapBackendFactory.kt [hunk 1]
```diff
@@ -35,7 +36,7 @@ class ImapBackendFactory(
     }
 
     private fun createImapStore(account: Account): ImapStore {
-        val oAuth2TokenProvider: OAuth2TokenProvider? = null
+        val oAuth2TokenProvider: OAuth2TokenProvider? = AndroidAccountOAuth2TokenStore(this.context)
```
Comment (wiktor-k) [inline app/k9mail/src/main/java/com/fsck/k9/backends/ImapBackendFactory.kt]: This should be injected I think...

File: app/ui/src/main/java/com/fsck/k9/activity/setup/AccountSetupBasics.java [hunk 2]
```diff
@@ -369,6 +376,21 @@ private void onNext() {
     @Override
     public void onActivityResult(int requestCode, int resultCode, Intent data) {
         if (resultCode == RESULT_OK) {
+            if (requestCode == CHOOSE_ACCOUNT) {
+                final String accountName = data.getStringExtra(AccountManager.KEY_ACCOUNT_NAME);
+                new AndroidAccountOAuth2TokenStore(this.getApplicationContext()).authorizeApi(accountName, this, new OAuth2TokenProvider.OAuth2TokenProviderAuthCallback() {
```
Comment (wiktor-k) [inline app/ui/src/main/java/com/fsck/k9/activity/setup/AccountSetupBasics.java]: This should also be either injected or `authorizeApi` should be converted to static...

File: app/ui/src/main/java/com/fsck/k9/activity/setup/AccountSetupBasics.java [hunk 3]
```diff
@@ -369,6 +376,21 @@ private void onNext() {
...
+                    @Override
+                    public void success() {
+                        onAccountManagerSetup(accountName);
+                    }
+
+                    @Override
+                    public void failure(AuthorizationException e) {
+                        Timber.e(e);
```
Comment (wiktor-k) [inline app/ui/src/main/java/com/fsck/k9/activity/setup/AccountSetupBasics.java]: There is no error recovery currently, if the user declines, they must go through manual setup or try again clicking `XOAUTH` button...

File: app/ui/src/main/java/com/fsck/k9/activity/setup/AccountSetupBasics.java [hunk 4]
```diff
@@ -423,15 +445,44 @@ private void onManualSetup() {
...
 
+    private void onAccountManagerSetup(String email) {
+        if (mAccount == null) {
+            mAccount = Preferences.getPreferences(this).newAccount();
+            mAccount.setChipColor(AccountCreator.pickColor(this));
+        }
+        mAccount.setName(getOwnerName());
+        mAccount.setEmail(email);
+
+        ServerSettings storeServer = new ServerSettings(Protocols.IMAP, "imap.gmail.com", -1,
```
Comment (wiktor-k) [inline app/ui/src/main/java/com/fsck/k9/activity/setup/AccountSetupBasics.java]: These server constants work well even for Google for Domains but maybe they should be moved somewhere else (to the OAuth provider?)

File: app/ui/src/main/java/com/fsck/k9/activity/setup/AccountSetupBasics.java [hunk 5]
```diff
@@ -423,15 +445,44 @@ private void onManualSetup() {
...
         } else if (id == R.id.manual_setup) {
             onManualSetup();
+        } else if (id == R.id.account_manager_setup) {
+            onAccountManagerSetup();
         }
     }
 
+    private void onAccountManagerSetup() {
+        Intent intent = AccountManager.newChooseAccountIntent(null, null,
+                new String[] { "com.google" }, true, null, null,
```
Comment (wiktor-k) [inline app/ui/src/main/java/com/fsck/k9/activity/setup/AccountSetupBasics.java]: This constant (or the entire intent creation?) should go to OAuthProvider...

File: mail/common/src/main/java/com/fsck/k9/mail/oauth/AndroidAccountOAuth2TokenStore.java [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+
+    @Override
+    public void authorizeApi(final String emailAddress, final Activity activity, final OAuth2TokenProviderAuthCallback callback) {
+        Account account = new Account(emailAddress, GOOGLE_ACCOUNT_TYPE);
+        accountManager.getAuthToken(account, GMAIL_AUTH_TOKEN_TYPE, null,
+                activity, new AccountManagerCallback<Bundle>() {
+                    @Override
+                    public void run(AccountManagerFuture<Bundle> accountManagerFuture) {
+                        try {
+                            accountManagerFuture.getResult();
```
Comment (wiktor-k) [inline mail/common/src/main/java/com/fsck/k9/mail/oauth/AndroidAccountOAuth2TokenStore.java]: Get result will throw with 3 kinds of exceptions, Philips code threw a detailed exception with info on what went wrong...

File: mail/common/src/main/java/com/fsck/k9/mail/oauth/AndroidAccountOAuth2TokenStore.java [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+            AccountManagerFuture<Bundle> future = accountManager
+                    .getAuthToken(account, GMAIL_AUTH_TOKEN_TYPE, false, null, null);
+            Bundle bundle = future.getResult(timeoutMillis, TimeUnit.MILLISECONDS);
+            if (bundle == null) {
+                throw new AuthenticationFailedException("No token provided");
+            }
+            if (bundle.get(AccountManager.KEY_ACCOUNT_NAME) == null) {
+                throw new AuthenticationFailedException("No account information provided");
+            }
+            if (bundle.get(AccountManager.KEY_ACCOUNT_NAME).equals(username)) {
```
Comment (wiktor-k) [inline mail/common/src/main/java/com/fsck/k9/mail/oauth/AndroidAccountOAuth2TokenStore.java]: I'm not entirely sure that these checks are important, if we ask for account `X` the result being for account `Y` seems unlikely...

File: mail/common/src/main/java/com/fsck/k9/mail/oauth/AndroidAccountOAuth2TokenStore.java [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+            fetchNewAuthToken(username, timeoutMillis);
+        }
+        return authTokens.get(username);
+    }
+
+    private void fetchNewAuthToken(String username, long timeoutMillis)
+            throws AuthenticationFailedException {
+        try {
+            Account account = new Account(username, GOOGLE_ACCOUNT_TYPE);
+            AccountManagerFuture<Bundle> future = accountManager
```
Comment (wiktor-k) [inline mail/common/src/main/java/com/fsck/k9/mail/oauth/AndroidAccountOAuth2TokenStore.java]: This is a deprecated method, I'll need to clean it up...

**Comment by morckx** [comment]
I haven't had a look at the code, but I can can confirm it's working well.

File: app/ui/src/main/java/com/fsck/k9/activity/setup/AccountSetupBasics.java (hunk 4 above)
Comment (philipwhiuk) [reply app/ui/src/main/java/com/fsck/k9/activity/setup/AccountSetupBasics.java]: Part of the complexity was working out how Google specific to make it - quite a few of the providers use XOAUTH2 now. Ideally there'd be no reference to Google anywhere except in a configuration system.

File: mail/common/src/main/java/com/fsck/k9/mail/oauth/AndroidAccountOAuth2TokenStore.java (hunk 7 above)
Comment (philipwhiuk) [reply mail/common/src/main/java/com/fsck/k9/mail/oauth/AndroidAccountOAuth2TokenStore.java]: Paranoia from seeing what some Android manufacturers manage to ship.

**Comment by philipwhiuk** [comment]
Thanks for picking it up!

File: app/ui/src/main/java/com/fsck/k9/activity/setup/AccountSetupBasics.java (hunk 4 above)
Comment (wiktor-k) [reply app/ui/src/main/java/com/fsck/k9/activity/setup/AccountSetupBasics.java]: Good idea, I've seen a ticket about Yahoo (#1698), do you know if it's working and how to test it? (If we need to register the app etc...)

Comment (philipwhiuk) [reply app/ui/src/main/java/com/fsck/k9/activity/setup/AccountSetupBasics.java]: We do, so it doesn't and it's basically impossible to get an API key - I tried tweeting a bigwig and Yahoo but it didn't go anywhere. 

I think getting it working with Microsoft might be doable.

However, the number of requests for those is much smaller. I'd try to ship Google and get feedback - it'll solve 90% of people's demands - you can always refactor later.

**Comment by wiktor-k** [comment]
@morckx Thanks for testing! I really appreciate it.

@philipwhiuk 
> I'd try to ship Google and get feedback - it'll solve 90% of people's demands - you can always refactor later.

Agreed. I try to make this PR as small as possible so it's easier to review and merge. Especially since #3661 some people can't use K-9 at all :-/

By the way would you mind if I added you as [`Co-authored-by`](https://blog.github.com/2018-01-29-commit-together-with-co-authors/)? Majority of the critical code here is taken from your super-old commits and just refactored a little bit...

**Comment by justinacolmena** [comment]
On December 8, 2018 11:57:14 AM AKST, Wiktor <notifications@github.com> wrote:
>@morckx Thanks for testing! I really appreciate it.
>
>@philipwhiuk 
>> I'd try to ship Google and get feedback - it'll solve 90% of people's
>demands - you can always refactor later.
>
>Agreed. I try to make this PR as small as possible so it's easier to
>review and merge. Especially since #3661 some people can't use K-9 at
>all :-/
>
>By the way would you mind if I added you as
>[`Co-authored-by`](https://blog.github.com/2018-01-29-commit-together-with-co-authors/)?
>Majority of the critical code here is taken from your super-old commits
>and just refactored a little bit...
>
>-- 
>You are receiving this because you are subscribed to this thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/k9mail/k-9/pull/3804#issuecomment-445489333


It looks like a lot of work you people are putting in.

Personally I am very happy with K-9 Mail as a basic standards-compliant client to connect to my own email server, which of course I have to spend a little bit of my own time and money to keep running, as well as my personal phone/internet connection, etc.

Internet bigwigs really cracked down on spam some years ago, and they have not let up. But Google, Yahoo, Microsoft et alia still make the big bucks advertising on the Internet. It's an eyeballs business. They want you to be looking at the ads, and make that snap decision to buy something. That is what they sell to their wholesale customers and advertisers.

It it not the top priority of Google, Yahoo or other big businesses to deliver standards-compliant e-mail to consumers without the ads. I am definitely not in favor of such a marketing-rich, delivery-poor, big-business, almost "SaaS" way of doing things, it is so proprietary, and I don't think most free/open source software people are, either, but just realize that big businesses have very strong profit motives and their goals do not always align very well with what the rest of us would otherwise consider reasonable.

-- 
A well regulated Militia, being necessary to the security of a free State, the right of the people to keep and bear Arms, shall not be infringed.

https://www.colmena.biz/~justina/justina.colmena.asc

**Comment by philipwhiuk** [comment]
@wiktor-k That's absolutely fine by me :+1:

**Comment by wiktor-k** [comment]
Okay, I've updated the UI with suggestions from @Valodim, and rebased on top of master.

All Gmail specific code is now in `AndroidAccountOAuth2TokenStore`, including server names etc. so in theory supporting another XOAUTH2 provider would need changes only there.

**Comment by Valodim** [comment]
If we are to merge this, we'll need to care about what happens on the settings screen, and during account import/export, too.

**Comment by wiktor-k** [comment]
> what happens on the settings screen

Do you mean that currently there is no indication in settings that XOAUTH2 is used? I was looking at the screen and thinking about adding new state to "authentication" drop down...

> and during account import/export, too

Yep, is there a way to trigger that currently from master branch? I'd like to test what happens there...

> We could also merge this, work from there, and worst case we'll hide the UI again before a release.

Yep, I'm already hiding UI for pre-M devices so it's one line of code. But maybe in the worst case provide something like an experimental option? I don't know if it's a good idea but I guess there are people that want this desperately and unused code tends to rot :)

**Comment by sivaraam** [comment]
Just wondering after seeing the screen shots in the description. Is it not possible to sign in using XOAuth when the Google account hasn't been signed in to the device?

**Comment by ikirudennis** [comment]
I received a message from Yahoo today saying that they recommend updating any apps that access email to use oauth. Is there anything I can do to help move this pull request along?

**Comment by wiktor-k** [comment]
Hi @ikirudennis, I actually tested this only on Gmail accounts (where it works very fine as I'm using it for months daily).

The setup is tightly coupled with Android accounts so if Yahoo is adding an account there then it should be relatively simple to add. If not, it may be more difficult. You can contact me via email if you'd like to discuss this further. 

As for merging it's @cketti that decides when or if the PR is of quality needed to be included in the master branch.

**Comment by r7l** [comment]
Just got a mail from Google that they are about to disable password access soon:

> Starting February 15, 2021, G Suite accounts will only allow access to apps using OAuth. Password-based access will no longer be supported.

> Access to LSAs will be turned off in two stages:
> 
>  **June 15, 2020** - Users who try to connect to an LSA for the first time will no longer be able to do so. This includes third-party apps that allow password-only access to Google calendars, contacts, and email via protocols such as CalDAV, CardDAV and IMAP. Users who have connected to LSAs prior to this date will be able to continue using them until usage of all LSAs is turned off.
> **February 15, 2021** - Access to LSAs will be turned off for all G Suite accounts.

I would love to keep using K9 as it still is one of the best options around.

**Comment by wiktor-k** [comment]
@r7l The same has been posted on Google blog: https://gsuiteupdates.googleblog.com/2019/12/less-secure-apps-oauth-google-username-password-incorrect.html

Note that I think this applies to GSuite accounts only, not for all Gmail accounts.

**Comment by r7l** [comment]
@wiktor-k It might GSuite for now but i don't see Google stopping there. The issue for me is that i am using Google for GSuite accounts for the most part due to work related setups on GSuite. But i don't want to register my phone on these accounts and that's why i am using K9 instead of the default mail client.

**Comment by wiktor-k** [comment]
If you can build K-9 you may want to try this PR. I try to keep it rebased on top of master and I'm using the same version for months with no major issues.

**Comment by r7l** [comment]
@wiktor-k Is there any chance to get this feature merged into master and have it published to the app store? I've never build an app for Android and i am not really planning of doing so on a long term basis. The app still works with IMAP for at least a couple more months.

If this is implemented already, it might be possible to get it out to public up until Gmail is closing the feature down.

**Comment by morckx** [comment]
> If you can build K-9 you may want to try this PR. I try to keep it rebased on top of master and I'm using the same version for months with no major issues.

Many thanks for the update / rebase! I'm also using your PR since day one – and I can bearely remember the bad times when Google would have locked my account every time I arrived at an unusual location :).

**Comment by n9** [comment]
@r7l 

> @wiktor-k Is there any chance to get this feature merged into master and have it published to the app store?

According to commits it seems that @cketti is maintainer.

**Comment by r7l** [comment]
@cketti is there any chance this gets merged into k-9?

**Comment by Blackbat007** [comment]
Just wondering whether this might be merged any time soon?

**Comment by sevaho** [comment]
Hi @wiktor-k 

I am trying out your PR but I don't quite get what to do with the client id got from Google Credentials.

I am getting this error when clicking on any gmail account in the popup.

```
02-07 08:19:44.208 19291 27959 W Auth    : [GetToken] GetToken failed with status code: UNREGISTERED_ON_API_CONSOLE
02-07 08:19:44.210 19291 27959 W GLSActivity: 
02-07 08:19:44.210 19291 27959 W GLSActivity: gyu: UNREGISTERED_ON_API_CONSOLE
02-07 08:19:44.210 19291 27959 W GLSActivity: 	at haw.getAuthToken(:com.google.android.gms@20104028@20.1.04 (100400-288960190):30)
02-07 08:19:44.210 19291 27959 W GLSActivity: 	at android.accounts.AbstractAccountAuthenticator$Transport.getAuthToken(AbstractAccountAuthenticator.java:247)
02-07 08:19:44.210 19291 27959 W GLSActivity: 	at android.accounts.IAccountAuthenticator$Stub.onTransact(IAccountAuthenticator.java:114)
02-07 08:19:44.210 19291 27959 W GLSActivity: 	at android.os.Binder.transact(Binder.java:688)
02-07 08:19:44.210 19291 27959 W GLSActivity: 	at duu.onTransact(:com.google.android.gms@20104028@20.1.04 (100400-288960190):2)
02-07 08:19:44.210 19291 27959 W GLSActivity: 	at android.os.Binder.transact(Binder.java:688)
02-07 08:19:44.210 19291 27959 W GLSActivity: 	at aaxe.onTransact(:com.google.android.gms@20104028@20.1.04 (100400-288960190):26)
02-07 08:19:44.210 19291 27959 W GLSActivity: 	at android.os.Binder.execTransact(Binder.java:752)
02-07 08:19:44.222 17693 17693 E AccountSetupBasics: com.fsck.k9.mail.oauth.AuthorizationException: Authorization failed
02-07 08:19:44.222 17693 17693 E AccountSetupBasics: 	at com.fsck.k9.mail.oauth.AndroidAccountOAuth2TokenStore$authorizeApi$1.run(AndroidAccountOAuth2TokenStore.kt:32)
02-07 08:19:44.222 17693 17693 E AccountSetupBasics: 	at android.accounts.AccountManager$17.run(AccountManager.java:2209)
02-07 08:19:44.222 17693 17693 E AccountSetupBasics: 	at android.os.Handler.handleCallback(Handler.java:873)
02-07 08:19:44.222 17693 17693 E AccountSetupBasics: 	at android.os.Handler.dispatchMessage(Handler.java:99)
02-07 08:19:44.222 17693 17693 E AccountSetupBasics: 	at android.os.Looper.loop(Looper.java:193)
02-07 08:19:44.222 17693 17693 E AccountSetupBasics: 	at android.app.ActivityThread.main(ActivityThread.java:6898)
02-07 08:19:44.222 17693 17693 E AccountSetupBasics: 	at java.lang.reflect.Method.invoke(Native Method)
02-07 08:19:44.222 17693 17693 E AccountSetupBasics: 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:537)
02-07 08:19:44.222 17693 17693 E AccountSetupBasics: 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
02-07 08:19:44.222 17693 17693 E AccountSetupBasics: Caused by: android.accounts.AuthenticatorException: UNREGISTERED_ON_API_CONSOLE
02-07 08:19:44.222 17693 17693 E AccountSetupBasics: 	at android.accounts.AccountManager.convertErrorToException(AccountManager.java:2541)
02-07 08:19:44.222 17693 17693 E AccountSetupBasics: 	at android.accounts.AccountManager.access$700(AccountManager.java:166)
02-07 08:19:44.222 17693 17693 E AccountSetupBasics: 	at android.accounts.AccountManager$AmsTask$Response.onError(AccountManager.java:2384)
02-07 08:19:44.222 17693 17693 E AccountSetupBasics: 	at android.accounts.IAccountManagerResponse$Stub.onTransact(IAccountManagerResponse.java:70)
02-07 08:19:44.222 17693 17693 E AccountSetupBasics: 	at android.os.Binder.execTransact(Binder.java:752)
```

Steps to reproduce:

```sh
git clone https://github.com/k9mail/k-9
git fetch origin 'pull/3804/head:xoauth' # should be xoath2 but the branch already existed ;)
git merge xoauth
./gradlew assembleDebug
```

# How I got client id from Goolge (didn't do anything with it)

```sh
vim ./app/k9mail/src/main/AndroidManifest.xml
```

-> package name is "com.fsck.k9"


```sh
keytool -list -v -keystore ~/.android/debug.keystore -alias androiddebugkey -storepass android -keypass android
```

-> copy sha1

Do I need to do anything with the client id that was generated from Google Credentials?

**Comment by wiktor-k** [comment]
@sevaho,

> package name is "com.fsck.k9"

The package name for the debug build is `com.fsck.k9.debug`. See:

```
$ ~/Android/Sdk/build-tools/29.0.2/aapt dump badging ./app/k9mail/build/outputs/apk/debug/k9mail-debug.apk | grep package
package: name='com.fsck.k9.debug' versionCode='27002' versionName='5.703-SNAPSHOT' compileSdkVersion='29' compileSdkVersionCodename='10'
```

**Comment by cketti** [comment]
I've changed my mind on using `AccountManager`. We'll have to go the browser route to be able to support providers other than Gmail.

**Pull request closed without merging**